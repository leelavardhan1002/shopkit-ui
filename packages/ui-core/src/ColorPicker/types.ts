export interface Color {
  id: string;
  label: string;
  value: string;
  borderColor?: string;
  selectedBorderColor?: string;
  disabled?: boolean;
}

export interface CoreProps {
  colors: Array<Color>;
  colorValue?: Color;
  defaultColorValue?: Color;
  onColorChange?: (color?: Color) => void;
  disabled?: boolean;
}

export interface AppearanceProps {
  variant?: 'square' | 'pill';
  size?: 'small' | 'medium' | 'large';
  showTooltip?: boolean;
  showLabels?: boolean;
  borderRadius?: number;
  maxVisible?: number;
  maxVisibleLabel?: string;
}

export interface ColorPickerProps extends CoreProps, AppearanceProps {
  className?: string;
  'aria-label'?: string;
}
