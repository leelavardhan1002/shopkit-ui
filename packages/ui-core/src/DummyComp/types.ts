export interface DummyCompProps {
    // Add your props here
}