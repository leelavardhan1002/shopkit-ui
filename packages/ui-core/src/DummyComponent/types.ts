export interface DummyComponentProps {
    // Add your props here
}