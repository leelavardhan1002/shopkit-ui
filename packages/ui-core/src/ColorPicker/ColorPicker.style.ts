import Box from '@mui/material/Box';
import { styled } from '@mui/material/styles';

export const StyledColorPicker = styled(Box)(({ theme }) => ({
  display: 'flex',
  flexWrap: 'wrap',
  gap: theme.spacing(1),
  padding: theme.spacing(1),
  '&:focus': {
    outline: 'none',
  },
}));

export const StyledLabel = styled('span')(({ theme }) => ({
  display: 'block',
  marginTop: theme.spacing(0.5),
  fontSize: theme.typography.caption.fontSize,
  color: theme.palette.text.secondary,
  textAlign: 'center',
}));

export const StyledColorWrapper = styled(Box)(() => ({
  display: 'flex',
  flexDirection: 'column',
  alignItems: 'center',
  justifyContent: 'center',
}));

export const ExpandLabel = styled(Box)(() => ({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  cursor: 'pointer',
  fontWeight: 500,
  fontSize: '14px',
  color: '#666',
  textDecoration: 'underline',
}));
