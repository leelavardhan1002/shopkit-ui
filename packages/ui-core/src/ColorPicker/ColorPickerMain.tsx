import React, { useState } from 'react';
import {
  ExpandLabel,
  StyledColorPicker,
  StyledColorWrapper,
  StyledLabel,
} from './ColorPicker.style';
import { ColorBlock } from './components/ColorBlock';
import { useColorPickerRefRegistry } from './contexts/ColorPickerRefRegistry.context';
import { ColorPickerProps } from './types';

export const ColorPickerMain: React.FC<ColorPickerProps> = (props) => {
  const {
    colors,
    disabled,
    showLabels,
    className,
    maxVisible,
    'aria-label': ariaLabel = 'Color picker',
    maxVisibleLabel,
  } = props;
  const [expanded, setExpanded] = useState(false);

  const { containerRef } = useColorPickerRefRegistry();

  const shouldLimit = maxVisible && maxVisible > 0 && colors.length > maxVisible;
  const visibleColors = shouldLimit && !expanded ? colors.slice(0, maxVisible) : colors;
  const hiddenCount = shouldLimit && !expanded ? colors.length - maxVisible : 0;

  return (
    <StyledColorPicker
      ref={containerRef}
      className={className}
      role="radiogroup"
      aria-label={ariaLabel}
      aria-disabled={disabled}
    >
      {visibleColors.map((color, index) => {
        return (
          <StyledColorWrapper key={color.id}>
            <ColorBlock color={color} index={index} />
            {showLabels && <StyledLabel>{color.label}</StyledLabel>}
          </StyledColorWrapper>
        );
      })}
      {hiddenCount > 0 && (
        <StyledColorWrapper>
          <ExpandLabel
            onClick={() => setExpanded(true)}
            role="button"
            tabIndex={0}
            onKeyDown={(e) => e.key === 'Enter' && setExpanded(true)}
          >
            {maxVisibleLabel || `+${hiddenCount} more`}
          </ExpandLabel>
        </StyledColorWrapper>
      )}
    </StyledColorPicker>
  );
};
