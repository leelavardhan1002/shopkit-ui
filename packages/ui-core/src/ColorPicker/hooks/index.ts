export { useColorSelection } from './useColorSelection';
export { useKeyboardNavigation } from './useKeyboardNavigation';
