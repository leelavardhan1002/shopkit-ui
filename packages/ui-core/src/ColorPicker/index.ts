export { ColorPicker } from './ColorPicker';
export { ColorBlock } from './components/ColorBlock';
export type { ColorPickerProps, Color } from './types';