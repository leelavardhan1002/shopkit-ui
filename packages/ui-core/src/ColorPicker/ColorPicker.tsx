import { ColorPickerMain } from './ColorPickerMain';
import { ColorPickerPropsProvider } from './contexts/ColorPickerProps.context';
import { ColorPickerRefRegistryProvider } from './contexts/ColorPickerRefRegistry.context';
import { ColorPickerStateProvider } from './contexts/ColorPickerState.context';
import { ColorPickerProps } from './types';

export const ColorPicker: React.FC<ColorPickerProps> = (props) => {
  return (
    <ColorPickerPropsProvider props={props}>
      <ColorPickerStateProvider>
        <ColorPickerRefRegistryProvider>
          <ColorPickerMain {...props} />
        </ColorPickerRefRegistryProvider>
      </ColorPickerStateProvider>
    </ColorPickerPropsProvider>
  );
};
