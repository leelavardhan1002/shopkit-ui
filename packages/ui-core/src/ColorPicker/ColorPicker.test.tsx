import { fireEvent, screen } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { renderWithTheme } from '../utils/testUtils/renderWithThemeAndProps';
import { ColorPicker } from './ColorPicker';
import { Color, ColorPickerProps } from './types';

const mockColors: Color[] = [
  { id: '1', label: 'Red', value: '#ff0000' },
  { id: '2', label: 'Blue', value: '#0000ff' },
  { id: '3', label: 'Green', value: '#00ff00' },
  { id: '4', label: 'Yellow', value: '#ffff00' },
  { id: '5', label: 'Purple', value: '#800080' },
  { id: '6', label: 'White', value: '#ffffff', borderColor: '#cccccc' },
  { id: '7', label: 'Black', value: '#000000', selectedBorderColor: '#ff0000' },
  { id: '8', label: 'Gradient', value: 'linear-gradient(to right, #ff0000, #0000ff)' },
  { id: '9', label: 'Image', value: 'url(https://example.com/pattern.png)' },
];

const colorsWithDisabled: Color[] = [
  { id: '1', label: 'Red', value: '#ff0000' },
  { id: '2', label: 'Blue', value: '#0000ff', disabled: true },
  { id: '3', label: 'Green', value: '#00ff00' },
];

const defaultProps: ColorPickerProps = {
  colors: mockColors,
  variant: 'square',
  size: 'medium',
};

const setupElement = (props: Partial<ColorPickerProps> = {}) => {
  return renderWithTheme(<ColorPicker {...defaultProps} {...props} />);
};

describe('ColorPicker', () => {
  it('matches snapshot', () => {
    const { asFragment } = setupElement();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders all colors', () => {
    setupElement();
    expect(screen.getAllByRole('radio')).toHaveLength(mockColors.length);
  });

  it('handles uncontrolled selection', async () => {
    const onColorChange = jest.fn();
    setupElement({ onColorChange });

    const firstColor = screen.getAllByRole('radio')[0] as HTMLElement;
    await userEvent.click(firstColor);

    expect(onColorChange).toHaveBeenCalledWith(mockColors[0]);
  });

  it('handles controlled selection', async () => {
    const onColorChange = jest.fn();
    setupElement({ colorValue: mockColors[0], onColorChange });

    const secondColor = screen.getAllByRole('radio')[1] as HTMLElement;
    await userEvent.click(secondColor);

    expect(onColorChange).toHaveBeenCalledWith(mockColors[1]);
  });

  it('updates controlled value when colorValue prop changes', () => {
    const { rerender } = renderWithTheme(
      <ColorPicker {...defaultProps} colorValue={mockColors[0]} />
    );
    expect(screen.getAllByRole('radio')[0]).toHaveAttribute('aria-checked', 'true');

    rerender(<ColorPicker {...defaultProps} colorValue={mockColors[1]} />);
    expect(screen.getAllByRole('radio')[1]).toHaveAttribute('aria-checked', 'true');
  });

  it('supports keyboard navigation with ArrowRight', async () => {
    setupElement({ defaultColorValue: mockColors[0] });

    const firstColor = screen.getAllByRole('radio')[0] as HTMLElement;
    firstColor.focus();

    fireEvent.keyDown(firstColor, { key: 'ArrowRight' });
    expect(screen.getAllByRole('radio')[1]).toHaveFocus();
  });

  it('supports keyboard navigation with ArrowDown', async () => {
    setupElement({ defaultColorValue: mockColors[0] });

    const firstColor = screen.getAllByRole('radio')[0] as HTMLElement;
    firstColor.focus();

    fireEvent.keyDown(firstColor, { key: 'ArrowDown' });
    expect(screen.getAllByRole('radio')[1]).toHaveFocus();
  });

  it('supports keyboard navigation with ArrowLeft', async () => {
    setupElement({ defaultColorValue: mockColors[1] });

    const secondColor = screen.getAllByRole('radio')[1] as HTMLElement;
    secondColor.focus();

    fireEvent.keyDown(secondColor, { key: 'ArrowLeft' });
    expect(screen.getAllByRole('radio')[0]).toHaveFocus();
  });

  it('supports keyboard navigation with ArrowUp', async () => {
    setupElement({ defaultColorValue: mockColors[1] });

    const secondColor = screen.getAllByRole('radio')[1] as HTMLElement;
    secondColor.focus();

    fireEvent.keyDown(secondColor, { key: 'ArrowUp' });
    expect(screen.getAllByRole('radio')[0]).toHaveFocus();
  });

  it('wraps around when navigating right from last color', async () => {
    setupElement({ defaultColorValue: mockColors[mockColors.length - 1] });

    const lastColor = screen.getAllByRole('radio')[mockColors.length - 1] as HTMLElement;
    lastColor.focus();

    fireEvent.keyDown(lastColor, { key: 'ArrowRight' });
    expect(screen.getAllByRole('radio')[0]).toHaveFocus();
  });

  it('wraps around when navigating left from first color', async () => {
    setupElement({ defaultColorValue: mockColors[0] });

    const firstColor = screen.getAllByRole('radio')[0] as HTMLElement;
    firstColor.focus();

    fireEvent.keyDown(firstColor, { key: 'ArrowLeft' });
    expect(screen.getAllByRole('radio')[mockColors.length - 1]).toHaveFocus();
  });

  it('navigates to first color with Home key', async () => {
    setupElement({ defaultColorValue: mockColors[2] });

    const thirdColor = screen.getAllByRole('radio')[2] as HTMLElement;
    thirdColor.focus();

    fireEvent.keyDown(thirdColor, { key: 'Home' });
    expect(screen.getAllByRole('radio')[0]).toHaveFocus();
  });

  it('navigates to last color with End key', async () => {
    setupElement({ defaultColorValue: mockColors[0] });

    const firstColor = screen.getAllByRole('radio')[0] as HTMLElement;
    firstColor.focus();

    fireEvent.keyDown(firstColor, { key: 'End' });
    expect(screen.getAllByRole('radio')[mockColors.length - 1]).toHaveFocus();
  });

  it('handles Enter key selection', async () => {
    const onColorChange = jest.fn();
    setupElement({ onColorChange });

    const firstColor = screen.getAllByRole('radio')[0] as HTMLElement;
    firstColor.focus();
    fireEvent.keyDown(firstColor, { key: 'Enter' });

    expect(onColorChange).toHaveBeenCalledWith(mockColors[0]);
  });

  it('handles Space key selection', async () => {
    const onColorChange = jest.fn();
    setupElement({ onColorChange });

    const firstColor = screen.getAllByRole('radio')[0] as HTMLElement;
    firstColor.focus();
    fireEvent.keyDown(firstColor, { key: ' ' });

    expect(onColorChange).toHaveBeenCalledWith(mockColors[0]);
  });

  it('ignores other keys', async () => {
    const onColorChange = jest.fn();
    setupElement({ onColorChange });

    const firstColor = screen.getAllByRole('radio')[0] as HTMLElement;
    firstColor.focus();
    fireEvent.keyDown(firstColor, { key: 'a' });

    expect(onColorChange).not.toHaveBeenCalled();
  });

  it('respects global disabled state', async () => {
    const onColorChange = jest.fn();
    setupElement({ disabled: true, onColorChange });

    const firstColor = screen.getAllByRole('radio')[0] as HTMLElement;
    await userEvent.click(firstColor);

    expect(onColorChange).not.toHaveBeenCalled();
  });

  it('respects individual color disabled state', async () => {
    const onColorChange = jest.fn();
    setupElement({ colors: colorsWithDisabled, onColorChange });

    const disabledColor = screen.getAllByRole('radio')[1] as HTMLElement;
    await userEvent.click(disabledColor);

    expect(onColorChange).not.toHaveBeenCalled();
  });

  it('renders with showLabels enabled', () => {
    setupElement({ showLabels: false });
    expect(screen.getAllByRole('radio')).toHaveLength(mockColors.length);
  });

  it('renders with showTooltip disabled', () => {
    setupElement({ showTooltip: false });
    expect(screen.getAllByRole('radio')[0]).toBeInTheDocument();
  });

  it('applies square variant styles', () => {
    setupElement({ variant: 'square' });
    expect(screen.getAllByRole('radio')[0]).toBeInTheDocument();
  });

  it('applies pill variant styles', () => {
    setupElement({ variant: 'pill' });
    expect(screen.getAllByRole('radio')[0]).toBeInTheDocument();
  });

  it('applies small size', () => {
    setupElement({ size: 'small' });
    expect(screen.getAllByRole('radio')[0]).toBeInTheDocument();
  });

  it('applies large size', () => {
    setupElement({ size: 'large' });
    expect(screen.getAllByRole('radio')[0]).toBeInTheDocument();
  });

  it('applies custom borderRadius', () => {
    setupElement({ borderRadius: 8 });
    expect(screen.getAllByRole('radio')[0]).toBeInTheDocument();
  });

  it('applies custom className', () => {
    setupElement({ className: 'custom-class' });
    expect(screen.getByRole('radiogroup')).toHaveClass('custom-class');
  });

  it('applies custom aria-label', () => {
    setupElement({ 'aria-label': 'Custom color picker' });
    expect(screen.getByRole('radiogroup')).toHaveAttribute('aria-label', 'Custom color picker');
  });

  it('uses default aria-label when not provided', () => {
    setupElement();
    expect(screen.getByRole('radiogroup')).toHaveAttribute('aria-label', 'Color picker');
  });

  it('limits visible colors with maxVisible', () => {
    setupElement({ maxVisible: 3 });
    expect(screen.getAllByRole('radio')).toHaveLength(3);
    expect(screen.getByText('+6 more')).toBeInTheDocument();
  });

  it('expands colors when clicking expand label', async () => {
    setupElement({ maxVisible: 3 });
    expect(screen.getAllByRole('radio')).toHaveLength(3);

    const expandLabel = screen.getByText('+6 more');
    await userEvent.click(expandLabel);

    expect(screen.getAllByRole('radio')).toHaveLength(mockColors.length);
  });

  it('expands colors when pressing Enter on expand label', async () => {
    setupElement({ maxVisible: 3 });
    expect(screen.getAllByRole('radio')).toHaveLength(3);

    const expandLabel = screen.getByText('+6 more');
    fireEvent.keyDown(expandLabel, { key: 'Enter' });

    expect(screen.getAllByRole('radio')).toHaveLength(mockColors.length);
  });

  it('uses custom maxVisibleLabel', () => {
    setupElement({ maxVisible: 3, maxVisibleLabel: 'Show all' });
    expect(screen.getByText('Show all')).toBeInTheDocument();
  });

  it('does not show expand label when maxVisible is 0', () => {
    setupElement({ maxVisible: 0 });
    expect(screen.getAllByRole('radio')).toHaveLength(mockColors.length);
  });

  it('does not show expand label when maxVisible is greater than colors length', () => {
    setupElement({ maxVisible: 20 });
    expect(screen.getAllByRole('radio')).toHaveLength(mockColors.length);
    expect(screen.queryByRole('button')).not.toBeInTheDocument();
  });

  it('sets correct tabIndex for first color when no selection', () => {
    setupElement();
    const firstColor = screen.getAllByRole('radio')[0] as HTMLElement;
    expect(firstColor).toHaveAttribute('tabIndex', '0');
  });

  it('sets correct tabIndex for selected color', () => {
    setupElement({ defaultColorValue: mockColors[2] });
    const selectedColor = screen.getAllByRole('radio')[2] as HTMLElement;
    expect(selectedColor).toHaveAttribute('tabIndex', '0');
  });

  it('sets correct tabIndex for disabled colors', () => {
    setupElement({ colors: colorsWithDisabled });
    const disabledColor = screen.getAllByRole('radio')[1] as HTMLElement;
    expect(disabledColor).toHaveAttribute('tabIndex', '-1');
  });

  it('handles colors with custom borderColor', () => {
    setupElement();
    const whiteColor = screen.getByLabelText('White');
    expect(whiteColor).toBeInTheDocument();
  });

  it('handles colors with custom selectedBorderColor', async () => {
    const onColorChange = jest.fn();
    setupElement({ onColorChange });
    
    const blackColor = screen.getByLabelText('Black');
    await userEvent.click(blackColor);
    
    expect(onColorChange).toHaveBeenCalledWith(mockColors[6]);
  });

  it('handles gradient colors', () => {
    setupElement();
    const gradientColor = screen.getByLabelText('Gradient');
    expect(gradientColor).toBeInTheDocument();
  });

  it('handles image colors', () => {
    setupElement();
    const imageColor = screen.getByLabelText('Image');
    expect(imageColor).toBeInTheDocument();
  });

  it('handles defaultColorValue in uncontrolled mode', () => {
    setupElement({ defaultColorValue: mockColors[1] });
    const secondColor = screen.getAllByRole('radio')[1] as HTMLElement;
    expect(secondColor).toHaveAttribute('aria-checked', 'true');
  });

  it('does not call onColorChange when disabled color is clicked', async () => {
    const onColorChange = jest.fn();
    setupElement({ colors: colorsWithDisabled, onColorChange });

    const disabledColor = screen.getAllByRole('radio')[1] as HTMLElement;
    await userEvent.click(disabledColor);

    expect(onColorChange).not.toHaveBeenCalled();
  });
});
