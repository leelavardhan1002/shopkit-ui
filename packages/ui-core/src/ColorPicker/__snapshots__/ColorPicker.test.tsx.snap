// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker matches snapshot 1`] = `
<DocumentFragment>
  <div
    aria-label="Color picker"
    class="MuiBox-root css-ve81fp"
    role="radiogroup"
  >
    <div
      class="MuiBox-root css-u63ya"
    >
      <div
        aria-checked="false"
        aria-label="Red"
        class="MuiBox-root css-1woxkw"
        role="radio"
        style="opacity: 1; cursor: pointer;"
        tabindex="0"
      />
    </div>
    <div
      class="MuiBox-root css-u63ya"
    >
      <div
        aria-checked="false"
        aria-label="Blue"
        class="MuiBox-root css-lxe1mo"
        role="radio"
        style="opacity: 1; cursor: pointer;"
        tabindex="-1"
      />
    </div>
    <div
      class="MuiBox-root css-u63ya"
    >
      <div
        aria-checked="false"
        aria-label="Green"
        class="MuiBox-root css-18xtvkq"
        role="radio"
        style="opacity: 1; cursor: pointer;"
        tabindex="-1"
      />
    </div>
    <div
      class="MuiBox-root css-u63ya"
    >
      <div
        aria-checked="false"
        aria-label="Yellow"
        class="MuiBox-root css-b6q2nl"
        role="radio"
        style="opacity: 1; cursor: pointer;"
        tabindex="-1"
      />
    </div>
    <div
      class="MuiBox-root css-u63ya"
    >
      <div
        aria-checked="false"
        aria-label="Purple"
        class="MuiBox-root css-196t0vj"
        role="radio"
        style="opacity: 1; cursor: pointer;"
        tabindex="-1"
      />
    </div>
    <div
      class="MuiBox-root css-u63ya"
    >
      <div
        aria-checked="false"
        aria-label="White"
        class="MuiBox-root css-1okyw7r"
        role="radio"
        style="opacity: 1; cursor: pointer;"
        tabindex="-1"
      />
    </div>
    <div
      class="MuiBox-root css-u63ya"
    >
      <div
        aria-checked="false"
        aria-label="Black"
        class="MuiBox-root css-hlo4lg"
        role="radio"
        style="opacity: 1; cursor: pointer;"
        tabindex="-1"
      />
    </div>
    <div
      class="MuiBox-root css-u63ya"
    >
      <div
        aria-checked="false"
        aria-label="Gradient"
        class="MuiBox-root css-tqw2l8"
        role="radio"
        style="opacity: 1; cursor: pointer;"
        tabindex="-1"
      />
    </div>
    <div
      class="MuiBox-root css-u63ya"
    >
      <div
        aria-checked="false"
        aria-label="Image"
        class="MuiBox-root css-1bd4eux"
        role="radio"
        style="opacity: 1; cursor: pointer;"
        tabindex="-1"
      />
    </div>
  </div>
</DocumentFragment>
`;
